src/mlpack/methods/neat/acyclic_net_impl.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * @file acyclic_net_impl.hpp
+ * @author Rahul Ganesh Prabhu
+ *
+ * Implementation of the AcyclicNet class, which represents an acyclic neural
+ * network.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_NEAT_ACYCLIC_NET_IMPL_HPP
+#define MLPACK_METHODS_NEAT_ACYCLIC_NET_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "acyclic_net.hpp"
+
+namespace mlpack{
+namespace neat /** NeuroEvolution of Augmenting Topologies */{
+
+// Creates an AcyclicNet object.
+template <class ActivationFunction>
+AcyclicNet<ActivationFunction>::AcyclicNet(const size_t nodeCount,
+                                           const size_t inputNodeCount,
+                                           const size_t outputNodeCount,
+                                           const double bias):
+    nodeCount(nodeCount),
+    inputNodeCount(inputNodeCount),
+    outputNodeCount(outputNodeCount),
+    bias(bias)
+{ /* Nothing to do here */ }
+
+// Evaluate a given input.
+template <class ActivationFunction>
+void AcyclicNet<ActivationFunction>::Evaluate(const arma::vec& input,
+                                              arma::vec& output,
+                                              const std::map<size_t, std::map
+                                                  <size_t, ConnectionGene>>&
+                                                  directedGraph,
+                                              const std::vector<size_t>&
+                                                  nodeDepths)
+{
+  arma::vec nodeValues(nodeCount, arma::fill::zeros);
+
+  // Populate the layers.
+  for (size_t i = 0; i < nodeCount; i++)
+  {
+    layers.reserve(nodeDepths[i] + 1);
+    while (layers.size() < nodeDepths[i] + 1)
+      layers.emplace_back(std::vector<size_t>());
+    layers[nodeDepths[i]].push_back(i);
+  }
+
+  nodeValues[0] = bias;
+
+  // Add all the nodes to the map.
+  for (size_t i = 1; i <= inputNodeCount; i++)
+    nodeValues[i] = input[i - 1];
+
+  // Activate the layers one by one.
+  for (size_t i = 0; i < layers.size(); i++)
+  {
+    for (size_t j = 0; j < layers[i].size(); j++)
+    {
+      size_t nodeID = layers[i][j];
+      // If this is a bias node, we need to add bias to it's connections.
+      if (nodeID == 0)
+      {
+        for (auto const& x : directedGraph[nodeID])
```

Comment:
We might have to change this to ``const auto const& x : directedGraph[nodeID]`` or I guess we could also use a standard for loop and use the index.

Reply:
This would give an error, since there would be two const qualifiers

Reply:
Ohh, right I missed to remove the second const.